### ISSUE TITLE
Port Tensor library from corefxlab to corefx

### ISSUE THREAD

**Issue by eerhardt** [dotnet/corefx#25779]

**Comment by karelz** [comment]
@eerhardt does it need to go through API review?

**Comment by eerhardt** [comment]
I think it already went through a first round, but yes, we should do a final/official API review.

**Comment by karelz** [comment]
Can you please post the API shape here? Or link to it if it is "too big". Then mark the issue as api-ready-for-review? Please ping Immo to schedule it, thanks!

**Comment by danmosemsft** [comment]
@tannergooding not sure whether you've already given feedback on this API, or whether it's the sort of API that interests you.

**Comment by tannergooding** [comment]
I've been following it (but not very closely).

The API shape overall looks reasonable to me.
  * nit: it would be helpful to add a newline between the logical API groupings (i.e. between `ToCompressedSparseTensor` and `ToSparseTensor` or between the constructors, properties, and methods, etc)

My feedback:
`ArrayTensorExtensions.ToTensor` should be `ArrayTensorExtensions.ToDenseTensor`, to match the naming of the other `To*Tensor` methods (and also to match `Tensor.ToDenseTensor`).

I would prefer if the `Clone` methods were explicit (i.e. `DeepClone` or `ShallowClone`)

Why `GetValue` and `SetValue` instead of an indexer?

Do we know if the C# language feature is going to be `Range` or `Range<T>`?

**Comment by terrajobst** [comment]
This looks like an API that we'll need 2 hours for.

@eerhardt @OliaG How soon does this need to happen?

**Comment by eerhardt** [comment]
> This looks like an API that we'll need 2 hours for.

Yes, it is a larger API and will take some time.  Especially the discussion around the Range type.

> How soon does this need to happen?

The plan is to ship this in .NET Core 2.1.  I'd like to get it into the first preview so it has more bake time, and more chance of getting feedback.

**Comment by eerhardt** [comment]
> nit: it would be helpful to add a newline between the logical API groupings (i.e. between ToCompressedSparseTensor and ToSparseTensor or between the constructors, properties, and methods, etc)

Good call.  Fixed.

> ArrayTensorExtensions.ToTensor should be ArrayTensorExtensions.ToDenseTensor, to match the naming of the other To*Tensor methods (and also to match Tensor.ToDenseTensor).

I had the same feedback, but @ericstj's thoughts were: https://github.com/dotnet/corefxlab/pull/1806#discussion_r142194947

> I would prefer if the Clone methods were explicit (i.e. DeepClone or ShallowClone)

I have no strong opinion either way.  I can let the API review team decide.  Here are my initial thoughts:
* The typical usage is to use numbers in a Tensor, which "Clone" may be explicit enough for numbers.
* ShallowCloneEmpty sounds odd to me.

> Why GetValue and SetValue instead of an indexer?

There are indexers for the n-dimensional indices.  GetValue and SetValue are for the linearized index, for the cases where a caller already has the linearized index and doesn't want/need to convert them into n-dimensional indices (which then get turned back into the linearized index under the covers).

> Do we know if the C# language feature is going to be Range or Range<T>?

I took the above proposal from [their proposal](https://github.com/dotnet/roslyn/blob/features/range/docs/features/range.md) (minus the IEnumerable part, since I don't need that).

**Comment by danmosemsft** [comment]
Presumably we need Roslyn to confirm the Range type, since they're going to adopt it. @khyperia

**Comment by tannergooding** [comment]
> I had the same feedback, but @ericstj's thoughts were: dotnet/corefxlab#1806 (comment)

I feel like that violates one of the .NET Naming Guidelines (https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/), but I don't have time to go searching through and finding it right now.

**Comment by terrajobst** [comment]
@tannergooding 

It doesn't violate the naming convention because ToTensor() does a return an instance of the `Tensor<T>` class -- it's just a more specific subclass.

**Comment by tannergooding** [comment]
@terrajobst: Having a method with a signature of `<SubType> To<Type>()`, at least to me, is not intuitive.

I would have expected the signature of an API called `ToTensor()` to be `Tensor<T> ToTensor()`. Then it would be returning `Tensor` as part of its public contract, but could internally return a different subtype and still be accurate.

Think of it this way: If `Tensor<T>` was not abstract, then `DenseTensor<T> ToTensor()` would prevent the existence of `Tensor<T> ToTensor()`, since overloading by return type isn't allowed in C#.

**Comment by danmosemsft** [comment]
@jaredpar @nguerrera FYI this proposes to depend on Range so we should coordinate on implementation.

**Comment by nietras** [comment]
Please see https://github.com/dotnet/corefxlab/issues/1938 for a critique of this API design.

**Comment by karelz** [comment]
API review comments:
* `CompressedTensor`
    * Indexer - changing recalculates the encoding - can we avoid that?
    * Change Memory-returning getters to ReadOnlyMemory (`CompressedCounts` and `Indices`)
* `Tensor.Slice` - does it have to be fast? Should we create `TensorView` struct?
* Deep copy vs. shallow copy and where we in general reuse underlying – should be clear from API name
* Arithmetic is troublesome - remove it for preview1
    * `Tensor<T>` should have `where T : IComparable<T>` instead
* `CloneEmpty` may need better name (Clone contradicts Empty)
* Linearized indexes - why are they exposed in public APIs? (`GetValue`, `SetValue`, `IndexOf`, `CopyTo`)
* `GetArrayString` rename to `ToString` (overriding object + overload with bool includeWhitespace)
* `Compare` is questionable (`IStructuralComparable`) - it may make sense on arrays, but unclear on multi-dimensional tensors

**Comment by karelz** [comment]
FYI:
The first API review discussion on 12/12 was recorded - see https://youtu.be/KKUjm8zaJBM?t=3014 (65 min duration)

The follow up 2h API review discussion on 12/14 was NOT recorded - see the notes above: https://github.com/dotnet/corefx/issues/25779#issuecomment-351883459

**Comment by jkotas** [comment]
> Make it easy for Machine Learning library vendors like CNTK, Tensorflow, Caffe, Scikit-Learn to port their libraries over

Have we proved that the proposed API achieves this goal?

It seems that this API will very likely need iterating on before it is declared stable, similar to how we have been iterating on Span/Memory or pipelines to make them work well for ASP.NET Core.

**Comment by nietras** [comment]
> It seems that this API will very likely need iterating on before it is declared stable, similar to how we have been iterating on Span/Memory or pipelines to make them work well for ASP.NET Core.

I very much agree with @jkotas that this proposal needs time to mature before making a proper proposal for the BCL. Not the least being real usage in different use cases. 

One also has to consider what the target area is for this? Is it for running inference in production? Is it for "off-line" learning? Real time learning? Ease of use? Each area would have their own requirements, 

I favor something that can in fact be used for inference in production with as low an overhead as possible, this means low level primitives with great performance is key, while ease of use is not the first priority. That can be build on top of that.

**Comment by helloguo** [comment]
How about `abs`,  `neg` and `pow`? Other tensor operations listed in following references could be interesting as well.

Ref:
https://www.cntk.ai/pythondocs/_modules/cntk/tensor.html
https://www.tensorflow.org/api_docs/python/tf/Tensor
https://github.com/onnx/onnx/blob/master/onnx/defs/math/defs.cc

**Comment by fuglede** [comment]
A mathematician's nitpick on naming: A [tensor](https://en.wikipedia.org/wiki/Tensor) is only [represented by a multi-dimensional array](https://en.wikipedia.org/wiki/Tensor#As_multidimensional_arrays) (which is what this data structure represents) after a choice of basis, and its representation will depend on that choice. You wouldn't have named `System.Numerics.Matrix` `System.Numerics.LinearTransformation`.

**Comment by danmosemsft** [comment]
@tannergooding can give an update here.
